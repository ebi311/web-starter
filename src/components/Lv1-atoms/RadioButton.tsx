import React, { useMemo, useCallback, ChangeEvent } from 'react';
import { styled } from '../Theme';

const Container = styled.div`
  display: flex;
  flex-direction: row;
`;

interface IProps {
  className?: string;
  name: string;
  selectItems: {
    label: string;
    value: string;
  }[];
  onChange?: (value: string) => void;
  value: string;
}

const RadioButton: React.FC<IProps> = props => {
  const onChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    if (!props.onChange) return;
    if (e.currentTarget.checked) {
      props.onChange(e.currentTarget.value);
    }
  }, []);
  const selectItems = useMemo(() => {
    return props.selectItems.map(i => (
      <label key={i.value}>
        <input
          type="radio"
          name={props.name}
          value={i.value}
          onChange={onChange}
          checked={props.value === i.value}
        />
        {i.label}
      </label>
    ));
  }, [props.selectItems, props.value]);
  return <Container className={props.className}>{selectItems}</Container>;
};

export default RadioButton;
