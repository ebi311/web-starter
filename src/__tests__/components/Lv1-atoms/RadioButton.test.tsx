import React from 'react';
import { shallow } from 'enzyme';
import RadioButton from '../../../components/Lv1-atoms/RadioButton';
describe('RadioButton', () => {
  const items = [
    {
      label: 'r01',
      value: 'v01',
    },
    {
      label: 'r02',
      value: 'v02',
    },
  ];
  test('success', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <RadioButton
        name="r001"
        selectItems={items}
        value="v01"
        onChange={onChange}
      />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
    const inputs = wrapper.find('input');
    expect(inputs).toHaveLength(2);
    const input01 = inputs.find('[value="v01"]');
    input01.simulate('change', {
      currentTarget: { checked: false, value: 'v02' },
    });
    expect(onChange).not.toBeCalled();
    const input02 = inputs.find('[value="v02"]');
    input02.simulate('change', {
      currentTarget: { checked: true, value: 'v02' },
    });
    expect(onChange).toBeCalledWith('v02');
  });
  test('success no onChange', () => {
    const wrapper = shallow(
      <RadioButton name="r001" selectItems={items} value="v01" />,
    ).dive();
    const input01 = wrapper.find('[value="v01"]');
    input01.simulate('change', {
      currentTarget: { checked: true, value: 'v02' },
    });
    // 例外が発生しなければOK
  });
});
