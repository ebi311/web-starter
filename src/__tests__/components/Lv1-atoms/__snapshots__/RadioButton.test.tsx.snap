// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton success 1`] = `
<div
  className="sc-AxjAm iXKhXH"
>
  <label
    key="v01"
  >
    <input
      checked={true}
      name="r001"
      onChange={[Function]}
      type="radio"
      value="v01"
    />
    r01
  </label>
  <label
    key="v02"
  >
    <input
      checked={false}
      name="r001"
      onChange={[Function]}
      type="radio"
      value="v02"
    />
    r02
  </label>
</div>
`;
